azure/services/dns/arecords.go

```diff
@@ -2,83 +2,137 @@ package dns
+					"DNSZone", s.scope.ClusterDomain(),
+					"hostname", desiredARecordSet.Hostname,
+				)
+				continue
+			}
+
+			aRecordsToCreate = append(aRecordsToCreate, desiredARecordSet)
+			logger.Info(
+				fmt.Sprintf("DNS A record '%s' is missing, it will be created", desiredARecordSet.Hostname),
+				"DNSZone", "",
+				"hostname", desiredARecordSet.Hostname)
+		}
+
+	}
+
+	return aRecordsToCreate
+
+}
+
+func (s *Service) updateARecords(ctx context.Context, currentRecordSets []*armdns.RecordSet) error {
+	logger := log.FromContext(ctx).WithName("arecords")
+	recordsToCreate := s.calculateMissingARecords(ctx, logger, currentRecordSets)
+
+	zoneName := s.scope.ClusterZoneName()
+
+	if len(recordsToCreate) == 0 {
+		logger.Info(
 			"All DNS A records have already been created",
-			"DNSZone", dnsSpec.ZoneName)
+			"DNSZone", zoneName)
 		return nil
 	}
 
-	for _, aRecord := range aRecords {
-		var ipAddressObject network.PublicIPAddress
-		ipAddressObject, err = s.publicIPsService.Get(ctx, s.Scope.ResourceGroup(), aRecord.PublicIPName)
+	for _, aRecord := range recordsToCreate {
+		ipAddressObject, err := s.publicIPsService.Get(ctx, &capzpublicips.PublicIPSpec{
+			Name:          aRecord.PublicIPName,
+			ResourceGroup: s.scope.ResourceGroup(),
+		})
 		if capzazure.ResourceNotFound(err) {
-			s.Scope.V(2).Info(
+			logger.Info(
 				"Cannot create DNS A record, public IP still not deployed",
-				"DNSZone", dnsSpec.ZoneName,
+				"DNSZone", zoneName,
 				"hostname", aRecord.Hostname,
 				"IP resource name", aRecord.PublicIPName)
 			continue
 		} else if err != nil {
 			return microerror.Mask(err)
 		}
 
-		if ipAddressObject.IPAddress == nil {
-			s.Scope.V(2).Info(
+		ipAddress := ipAddressObject.(network.PublicIPAddress).IPAddress
+
+		if ipAddress == nil {
+			logger.Info(
 				"Cannot create DNS A record, public Azure IP object does not have IP address set",
-				"DNSZone", dnsSpec.ZoneName,
+				"DNSZone", zoneName,
 				"hostname", aRecord.Hostname,
 				"IP resource name", aRecord.PublicIPName)
 			continue
 		}
 
-		s.Scope.V(2).Info(
+		logger.Info(
 			"Creating DNS A record",
-			"DNSZone", dnsSpec.ZoneName,
+			"DNSZone", zoneName,
 			"hostname", aRecord.Hostname,
-			"ipv4", *ipAddressObject.IPAddress)
+			"ipv4", *ipAddress)
 
-		recordSet := dns.RecordSet{
-			Type: to.StringPtr(string(dns.A)),
-			RecordSetProperties: &dns.RecordSetProperties{
-				ARecords: &[]dns.ARecord{
+		recordSet := armdns.RecordSet{
+			Type: to.StringPtr(string(armdns.RecordTypeA)),
+			Properties: &armdns.RecordSetProperties{
+				ARecords: []*armdns.ARecord{
 					{
-						Ipv4Address: ipAddressObject.IPAddress,
+						IPv4Address: ipAddress,
 					},
 				},
 				TTL: to.Int64Ptr(aRecord.TTL),
 			},
 		}
-		_, err := s.client.CreateOrUpdateRecordSet(
+		_, err = s.azureClient.CreateOrUpdateRecordSet(
 			ctx,
-			s.Scope.ResourceGroup(),
-			dnsSpec.ZoneName,
-			dns.A,
+			s.scope.ResourceGroup(),
+			s.scope.ClusterZoneName(),
```

Comment:
This is where the code is currently failing because it's trying to propagate the NS records to `/subscriptions/<sub-id>/resourceGroups/<cluster-name>/providers/Microsoft.Network/dnsZones/<base-domain>/NS/@` and the base zone is not created in that resource group. It is per customer and should be created in some universal resource group that this codebase is somehow aware of.